constph.py

```diff
@@ -732,7 +736,103 @@ def setNumAttemptsPerUpdate(self, nattempts=None):
+        """
+        target_weights (list) - None or nested list indexed[group][state] of relative weights (pi) for sams method
+        """
+
+        super(CalibrationTitration, self).__init__(system, temperature, pH, prmtop, cpin_filename, nattempts_per_update=None, simultaneous_proposal_probability=0.1, debug=False)
+
+        self.n_adaptations=0
+
+        for i,group in enumerate(self.titrationGroups):
+            for j, state in enumerate(self.titrationGroups[i]['titration_states']):
+                if target_weights is not None:
+                    self.titrationGroups[i]['titration_states'][j]['target_weight'] = target_weights[i][j]
+                else:
+                    self.titrationGroups[i]['titration_states'][j]['target_weight'] = 1.0 / len(self.titrationGroups[i]['titration_states'])
+
+
+    def adapt_weights(self, context, scheme):
+        self.n_adaptations += 1
+        temperature = self.temperature
+        kT = kB * temperature  # thermal energy
+        beta = 1.0 / kT  # inverse temperature
+        # zeta^{t-1}
+        zeta = self._get_zeta(beta)
+
+        if scheme in ['theorem1', 'eq9']:
+            update = self._theorem1()
+        elif scheme in ['theorem2', 'eq12']:
+            update = self._theorem2(context, beta, zeta)
+        else:
+            raise ValueError("Unknown adaptation scheme!")
+
+        # zeta^{t-1/2}
+        zeta += update
+        # zeta^{t} = zeta^{t-1/2} - zeta_1^{t-1/2}
+        zeta_t = zeta - zeta[0]
+
+        # Set reference energy based on new zeta
+        for i, titr_state in enumerate(zeta_t):
+            self.titrationGroups[0]['titration_states'][i]['relative_energy'] = titr_state / -beta
+
+    def _get_zeta(self, beta):
+        zeta = np.asarray(
+            map(lambda x: np.float64(x['relative_energy'] * -beta), self.titrationGroups[0]['titration_states'][:]))
+        return zeta
+
+    def _theorem1(self):
+        # [1/pi_1...1/pi_i]
+        update = np.asarray(map(lambda x: 1 / x['target_weight'], self.titrationGroups[0]['titration_states'][:]))
+        # delta(Lt)
+        delta = np.zeros_like(update)
+        delta[self.getTitrationState(0)] = 1
+        update *= delta
+        update /= self.n_adaptations  # t^{-1}
+        return update
+
+    def _theorem2(self, context, beta, zeta):
+        # target weights
+        pi_j = self._get_target_weights()
+        # [1/pi_1...1/pi_i]
+        update = np.apply_along_axis(lambda x: 1/x, 0, pi_j)
+        ub_j = self._get_potential_energies(beta, context, update)
+        # w_j(X;zeta)
+        log_w_j = np.log(pi_j) - zeta -ub_j
+        log_w_j -= logsumexp(log_w_j)
+        w_j = np.exp(log_w_j)
+        update *= w_j
+        update /= self.n_adaptations  # t^{-1}
+
+        return update
+
+    def _get_target_weights(self):
+        return np.asarray(map(lambda x: x['target_weight'], self.titrationGroups[0]['titration_states'][:]))
+
+    def _get_potential_energies(self, beta, context, update):
```

Comment:
Since this is a constant-pH simulation, q_j(x) = exp[-u_j(x)], but the reduced potential u_j(x) is not just beta*potential_energy.  Recall that this is a semigrand canonical ensemble, so we need to include the chemical potential term too into u_j(x) that depends on the pH and number of protons.